### DISCUSSION TITLE
Not able to read request headers in API middleware.ts

### DISCUSSION THREAD

**Discussion by dmudro** [vercel/next.js#44905] [Help] [answered]
I can't get a basic API [middleware auth example](https://nextjs.org/docs/advanced-features/middleware#producing-a-response) working in `middleware.ts`; it's almost copy paste from the doc example but does not compile in `13.1.2`. perhaps I am overlooking something obvious?

```
// middleware.ts
import { NextRequest, NextResponse } from 'next/server';
import { isAuthenticated } from '@lib/auth';

export const config = {
  matcher: '/api/:function*',
};

export function middleware(request: NextRequest) {
  console.log(request.headers); // TS ERROR: unknown property headers; cannot access directly or via `headers.get()`;

  if (!isAuthenticated(request)) {
    return new NextResponse( // tried `NextResponse.json() too
      JSON.stringify({ success: false, message: 'authentication failed' }),
      { status: 401, headers: { 'content-type': 'application/json' } } // TS ERROR: ts(ts2345); the signature does not recognise `status` or `headers`
    );
  }
}
```

**Comment by icyJoseph** [comment]
Works without issues for me.

- Are you sure you've got next@13.1.2? Could you run `yarn why next` or `npm explain next` in your terminal?
- Delete `node_modules` and `.next`, and install everything again.

Assuming you've got VSCode:

- Reload the VSCode instance (just close it and reopen or do `cmd+shift+p` window reload)
- Is the TypeScript version in your project the same as in the VSCode? Maybe your VSCode version is too low?

**Reply by dmudro** [reply]
thanks, good suggestions. no change after wiping `node_modules` & `.next` though while `npm explain next` reports

```
next@13.1.2
node_modules/next
  next@"^13.1.2" from the root project
```

![image](https://user-images.githubusercontent.com/1648143/212698108-9cb1aca4-599d-4ff1-92bb-75c5d5080133.png)

**Reply by icyJoseph** [reply]
Why are you running `tsc --noEmit` though?

**Reply by dmudro** [reply]
to copy paste the ts error / output here

**Reply by icyJoseph** [reply]
This is really strange then. Does next dev or next build work at all?

**Reply by dmudro** [reply]
silencing TS checker with `// @ts-ingore` compiles (obviously) and my middleware logic works as expected.

~~given the doc examples are plain JS~~ and some features in API middleware are experimental(?), I assume this is a type definition bug in `next/server`.

**Comment by dmudro** [comment]
same type error in Vercel. assuming this is not a code issue but incorrectly defined type in `next/server`. raising a bug soon.

![image](https://user-images.githubusercontent.com/1648143/212952488-0b68645f-a5c6-4bff-b3fd-f7f0f79bbe18.png)

**Comment by dmudro** [comment, marked as answer]
issue found: `tsconfig.json` was lacking `dom` in the `compilerOptions.lib` array. probably related to `Request` object / type.

```
// tsconfig.json
{
  "compilerOptions": {
    "lib": ["dom", "dom.iterable", "esnext"],
...
```

**Reply by icyJoseph** [reply]
Ah yes! wow good catch!
